compiler/rustc_mir_transform/src/shim/async_destructor_ctor.rs

```diff
@@ -121,9 +121,10 @@ pub(super) fn build_async_drop_shim<'tcx>(
         parent_args.as_coroutine().resume_ty(),
     )));
     body.phase = MirPhase::Runtime(RuntimePhase::Initial);
-    if !needs_async_drop {
+    if !needs_async_drop || matches!(drop_ty.kind(), ty::Error(_)) {
```

Comment:
maybe use is_tainted_by_errors on `drop_ty`, as it may contain drop things within. E.g. I think this may still ICE on `Struct<str>` being used for a `Struct<impl Sized>` return type

Reply:
Could you be more specific, pls?
I didn't find `is_tainted_by_errors` as a Ty function. `tainted_by_errors` is greppable, but looks like it is `body` flag, not bound to specific type. I am not sure if this flag will be set to the just-created shim body.

Reply:
Ah sorry, it's actually `references_error` from `TypeVisitableExt` which you may need to import.